org.lflang/src/org/lflang/generator/c/CGenerator.xtend

```diff
@@ -916,6 +922,50 @@ class CGenerator extends GeneratorBase {
         JavaGeneratorUtils.refreshProject(resource, context.mode)
     }
 
+    /**
+     * Add files needed for the proper function of the runtime scheduler to
+     * {@code coreFiles} and {@link TargetConfig#compileAdditionalSources}.
+     */
+    def addSchedulerFiles(ArrayList<String> coreFiles) {
+        coreFiles.add("threaded/scheduler.h")
+        coreFiles.add("threaded/scheduler_instance.h")
+        coreFiles.add("threaded/scheduler_sync_tag_advance.c")
+        // Don't use a scheduler that does not prioritize reactions based on deadlines
+        // if the program contains a deadline (handler). Use the GEDF_NP scheduler instead.
+        if (!targetConfig.schedulerType.prioritizesDeadline) {
+            // Check if a deadline is assigned to any reaction
+            if (reactors.filter[reactor |
+                // Filter reactors that contain at least one reaction 
+                // that has a deadline handler.
+                return reactor.allReactions.filter[ reaction |
+                    return reaction.deadline !== null
+                ].size > 0;
+            ].size > 0) {
+                
+                if (targetConfig.setByUser.contains(TargetProperty.SCHEDULER)) { 
+                    // FIXME: This potentially belongs in the validator
+                    // TODO: Move after LFValidator is ported
+                    errorReporter.reportWarning("This program contains deadlines, but the chosen "
+                        +targetConfig.schedulerType.name
+                        +" scheduler does not prioritize reaction execution "
+                        +"based on deadlines. This might result in a sub-optimal "
+                        +"scheduling.");
+                } else {
+                    targetConfig.schedulerType = TargetProperty.SchedulerOption.GEDF_NP;
```

Comment:
That's certainly doable. Should we create an issue for this or try to address it first before the merge?

Reply:
Perhaps we should defer this until https://github.com/lf-lang/lingua-franca/pull/971 has been merged, since it provides functionality for copying entire directories. I think "bloat" in `src-gen` isn't really really a concern; it won't affect the size of the produced binary...